text-document.rkt

```diff
@@ -567,24 +532,57 @@
-     (define range
+     (define-values (start end)
        (match ch
          ["\n"
-          (Range
-           #:start (abs-pos->Pos doc-text (send doc-text paragraph-start-position line))
-           #:end (abs-pos->Pos doc-text (send doc-text paragraph-end-position line)))]
+          (values
+           (abs-pos->Pos doc-text (send doc-text paragraph-start-position line))
+           (abs-pos->Pos doc-text (send doc-text paragraph-end-position line)))]
          [_
-          (Range
-           #:start (abs-pos->Pos doc-text (or (find-containing-paren pos (send doc-text get-text))
-                                              0))
-           #:end (abs-pos->Pos doc-text pos))]))
-
-     (range-formatting!
-      id
-      (hash-set params
-                'range range))]
+          (values
+           (abs-pos->Pos doc-text (or (find-containing-paren pos (send doc-text get-text))
+                                      0))
+           (abs-pos->Pos doc-text pos))]))
+     (success-response id (format! uri start end 'on-type))]
     [_
-     (error-response id INVALID-PARAMS "textDocument/onTypeformatting failed")]))
+     (error-response id INVALID-PARAMS "textDocument/onTypeFormatting failed")]))
+
+;; Shared path for all formatting requests
+;; `kind` allows diverging in a few places while sharing the core operation
+(define (format! uri start end kind)
+  (unless (uri-is-path? uri)
+    (error 'format! "uri is not a path"))
+  (match-define (doc doc-text doc-trace)
+    (hash-ref open-docs (string->symbol uri)))
+  (define indenter (send doc-trace get-indenter))
+  (define start-pos (Pos->abs-pos doc-text start))
+  ;; Adjust for line endings (#92)
+  (define end-pos (max start-pos (sub1 (Pos->abs-pos doc-text end))))
+  (define start-line (send doc-text position-paragraph start-pos))
+  (define end-line (send doc-text position-paragraph end-pos))
+  (define mut-doc-text
+    (if (is-a? doc-text racket:text%)
+        (let ([r-text (new racket:text%)])
+          (send r-text insert (send doc-text get-text))
+          r-text)
+        (send doc-text copy-self)))
+  (define skip-this-line? #f)
+  (if (eq? indenter 'missing) (json-null)
+      (let loop ([line start-line])
+        (define line-start (send mut-doc-text paragraph-start-position line))
+        (define line-end (send mut-doc-text paragraph-end-position line))
+        (for ([i (range line-start (add1 line-end))])
+          (when (and (char=? #\" (send mut-doc-text get-character i))
+                      (not (char=? #\\ (send mut-doc-text get-character (sub1 i)))))
+            (set! skip-this-line? (not skip-this-line?))))
+        (if (> line end-line)
+            null
+            (append (filter-map
+                      identity
+                      ; NOTE: The order is important somehow
+                      (list (remove-trailing-space! mut-doc-text skip-this-line? line)
+                            (indent-line! mut-doc-text indenter line kind)))
```

Comment:
The order is important, since remove space is "remove" the position. Consider this case:
```
xxx•••
indent to ~>
  xxx•••
then remove space cut off position ~>
  x••
```

Reply:
Right, it makes sense to me... The comment above is just your own comment from https://github.com/jeapostrophe/racket-langserver/commit/ad7c0c958968d526c1b27c4d40959d737fd81111 that I copied while extracting a shared path. 😄

Should we change the comment to explain the details or perhaps remove it?

Reply:
Oh! Did notice that from myself, but I think this is a good chance to add this into comment.

Reply:
Sounds good, I've added this detail in a longer comment.